### ISSUE TITLE
[bug] trellis alias causing bud build to log warning

### ISSUE THREAD

**Issue by chrillep** [roots/bud#2332]
### Agreement

- [X] This is not a duplicate of an existing issue
- [X] I have read the [guidelines for Contributing to Roots Projects](https://github.com/roots/.github/blob/master/CONTRIBUTING.md)
- [X] This is not a personal support request that should be posted on the [Roots Discourse](https://discourse.roots.io/) community

### Describe the issue

when trellis aliases are generated bud tries to read the alias file and throws a warning

relates to: https://github.com/roots/trellis-cli/issues/397

sidenote: does not throw warning in 
* v6.7.1
* v6.11.0

### Expected Behavior

idk should it read trellis alias files

### Actual Behavior

not to read trellis wp alias files?

### Steps To Reproduce

1. run trellis aliases
2. run yarn build or any bud command

### version

6.12.0

### Logs

```zsh
[fs] [wp-cli.trellis-alias.yml] › ⚠  wp-cli.trellis-alias.yml causes an exception when read.
Since wp-cli.trellis-alias.yml does not appear to be a bud configuration file, bud.js is not throwing. Original error follows:

BudError: end of the stream or a document separator is expected (1:1)

 1 | @development:
-----^
```

**Comment by kellymears** [comment]
`@` is technically valid yml but there isn't a json parser available which handles it without throwing.

For the moment bud.js will read or otherwise try to import any `yml`, `json`, `ts`, `js` files in the root of the project or in the `./config` directory, _except_ yarn, pnpm, & npm lockfiles as well as both the composer lockfile and manifest.

Eventually, bud.js will import/read project config files lazily, and this should stop being a problem then.

For now you can just quote the key and the warning will go away. It doesn't throw, it's a handled exception, and doesn't effect the build either way.

```yml
"@development":
  ssh: vagrant@foo/srv/www/foo/current
# ...etc
```

**Comment by chrillep** [comment]
> `@` is technically valid yml but there isn't a json parser available which handles it without throwing.
> 
> For the moment bud.js will read or otherwise try to import any `yml`, `json`, `ts`, `js` files in the root of the project or in the `./config` directory, _except_ yarn, pnpm, & npm lockfiles as well as both the composer lockfile and manifest.
> 
> Eventually, bud.js will import/read project config files lazily, and this should stop being a problem then.
> 
> For now you can just quote the key and the warning will go away. It doesn't throw, it's a handled exception, and doesn't effect the build either way.
> 
> ```yaml
> "@development":
>   ssh: vagrant@foo/srv/www/foo/current
> # ...etc
> ```
Really good to know! :) 

Should the trellis cli command `trellis alias` that generates the alias file quote the strings?
instead of every time alias files are generated, userland needs to edit them.

**Comment by swalkinshaw** [comment]
> bud.js will read or otherwise try to import any `yml`, `json`, `ts`, `js` files in the root of the project

@kellymears this seems weird... Can't there be a hard-coded list of files for Bud to attempt to load? Very odd that some random malformed config file could result in Bud throwing an error message.

**Comment by kellymears** [comment]
It doesn't throw. It's a log message prettily formatted as a warning and emitted to stdout from a catch block. The process continues and (if all is OK) exits 0.

I think the easiest fix is to just remove the log event, but I'm a little torn because this same message has helped me on a couple occasions. 

And, its pretty par for the course (in node land, at least). How many warnings are displayed during the average npm or yarn install?

**Comment by swalkinshaw** [comment]
That's all fair, but what's the purpose of trying to load _any_ yml, json, ts, js files?

**Comment by kellymears** [comment]
For the most part files in `./` or `./config` with those extensions _are_ configuration files which are relevant to the build. 

bud.js inspects the fs in order to generate a hash which is used as a checksum for versioning the filesystem cache (both bud's internal cache and the compiler's). This system was born out of necessity -- while `eslint-webpack-plugin` has gotten better about handling cache in recent updates, as an example, it used to only fail a build if the cache was empty. 

Since file IO performance is not a particular highlight of v8, why not also just store these files in memory for the duration of the process? Then, later on, an extension like stylelint can just grab the config module from `bud.context.files['stylelint.config'].module`, rather than ad hoc opening and manipulating another Buffer?

Two other bonuses are conferred by this:

- All these file ops can happen in parallel without needing to worry about dependency graphs or race conditions later on
- A config module can be authored in any of those formats regardless of whether the tool would accept that format normally. `stylelint.config.yml`? To be honest, I'm not sure if stylelint supports that, but I know that bud.js does.

Again, though, I think the ideal version of the system is to load files lazily (despite the advantages listed above). It's just an annoying refactor since every bud extension that wants access to the filesystem assumes that the work is already done and is relying on synchronous access to a static object. I have worked on it locally, but so far haven't been able to reach performance parity with what is released.

**Comment by kellymears** [comment]
This PR should resolve the issue (features lazy loaded config access): https://github.com/roots/bud/pull/2336

Still uses checksums to bust v8 cache and/or require re-compilation by esbuild-wasm (for ts config files), just doesn't perform op until requested. Little to no observable overhead.

**Comment by chrillep** [comment]
is there a release for this? How do i test it? 😅

**Comment by chrillep** [comment]
> This PR should resolve the issue (features lazy loaded config access): #2336
> 
> Still uses checksums to bust v8 cache and/or require re-compilation by esbuild-wasm (for ts config files), just doesn't perform op until requested. Little to no observable overhead.

Gr8! is there a release for this? How do i test it? 😅

**Comment by kellymears** [comment]
It'll be released as 6.14.0 when it does release.

I merged the PR earlier today so a new nightly should be automatically released tonight. 

@swalkinshaw I think I was wrong about `@` being fine to start a key (discovered when checking behavior of the above PR).

> The “@” (x40, at) and “`” (x60, grave accent) are reserved for future use.
> Reserved indicators can't start a plain scalar.
[yaml spec (1.2.2)](https://yaml.org/spec/1.2.2/)

`yamllint` and vscode report syntax errors, so I think my reading of this is correct.

```txt
test.yml
  1:1       error    syntax error: found character '@' that cannot start any token (syntax)
```

I opened a PR in case I'm right: https://github.com/roots/trellis-cli/pull/398

**Comment by kellymears** [comment]
In [6.14.1](https://github.com/roots/bud/releases/tag/v6.14.1) config modules are now lazy loaded. There shouldn't be a problem anymore unless there is an attempt to access `wp-cli` as a module (none of the code I have written does). In that case it wouldn't present a warning, it would throw a proper error.

Tested with this horrifying `wp-cli.yml` in the project directory:

```yml
@development:
foo

=:
  - foo!
  b? > foo
```

No problem, unless I try to get at it:

```ts
export default async bud => {
  await app.context.files[`wp-cli`].module();
}
```

Which throws:

```
ConfigurationError

✘ Error in bud.config

BudError

✘ end of the stream or a document separator is expected (1:1)

 1 | @development:
-----^
 2 | foo
 3 |

☰ Stack trace
BudError: end of the stream or a document separator is expected (1:1)
 1 | @development:
-----^
 2 | foo
 3 |
    at generateError (node_modules/@roots/filesystem/node_modules/js-yaml/dist/js-yaml.mjs:1273:10)
    at throwError (node_modules/@roots/filesystem/node_modules/js-yaml/dist/js-yaml.mjs:1277:9)
    at readDocument (node_modules/@roots/filesystem/node_modules/js-yaml/dist/js-yaml.mjs:2735:5)
    at loadDocuments (node_modules/@roots/filesystem/node_modules/js-yaml/dist/js-yaml.mjs:2778:5)
    at Object.load$1 [as load] (node_modules/@roots/filesystem/node_modules/js-yaml/dist/js-yaml.mjs:2804:19)
    at Module.read (node_modules/@roots/filesystem/lib/yml.js:5:17)
    at async Filesystem.read (node_modules/@roots/filesystem/lib/filesystem.js:164:20)
    at async file.module (node_modules/@roots/bud-support/lib/utilities/files.js:82:35)
    at async default (bud.config.js?v=8a08feab56f68ceab7d56c2914b9117b8a8bf0be:70:3)
    at async Configuration.dynamicConfig (node_modules/@roots/bud-framework/lib/configuration/configuration.js:25:20)

ℹ See file /******/******/*****/bud.config.js
```
